lwrunner/awsrunner.go

```diff
@@ -144,3 +113,65 @@ func (run AWSRunner) SendPublicKey(pubBytes []byte) error {
 
 	return nil
 }
+
+// getAMIName takes an AMI image ID and an AWS region name as input
+// and calls the AWS API to get the name of the AMI. Returns the AMI
+// name or an error if unsuccessful.
+func getAMIName(amiImageId, region string) (string, error) {
+	cfg, err := config.LoadDefaultConfig(context.Background())
+	if err != nil {
+		return "", err
+	}
+	cfg.Region = region
+	svc := ec2.NewFromConfig(cfg)
+	input := ec2.DescribeImagesInput{
+		ImageIds: []string{
+			amiImageId,
+		},
+	}
+	result, err := svc.DescribeImages(context.Background(), &input)
+	if err != nil {
+		return "", err
+	}
+	if len(result.Images) != 1 {
+		return "", fmt.Errorf("expected to find only one AMI, instead found %v", result.Images)
+	}
+
+	return *result.Images[0].Name, nil
+}
+
+// getSSHUsername takes any username passed as a CLI arg,
+// an AMI image name, a shell environment, and returns
+// the username for SSHing into the AWS runner or the empty
+// string and an error if the AMI is not supported.
+// It first checks if `LW_SSH_USER` is set and returns it if so.
+// Then it checks the AMI image name to heuristically determine the
+// SSH username.
+func getSSHUsername(userFromCLIArg, imageName string) (string, error) {
+	if userFromCLIArg != "" { // from CLI arg
+		return userFromCLIArg, nil
+	}
+	usernameLUT := getSSHUsernameLookupTable()
+	for _, matchFn := range usernameLUT {
+		if match, foundName := matchFn(imageName); match {
+			return foundName, nil
+		}
+	}
+	// No matching AMI found, return an error
+	return "", fmt.Errorf("no SSH username found for AMI %s, set as arg or shell env", imageName)
+}
+
+// getSSHUsernameLookupTable returns a lookup table for heuristically
+// determining SSH username based on AMI.
+// The first row of the table it returns is a function that checks
+// `LW_SSH_USER` in the shell environment.
+func getSSHUsernameLookupTable() []func(string) (bool, string) {
+	return []func(string) (bool, string){
+		func(_ string) (bool, string) { return os.Getenv("LW_SSH_USER") != "", os.Getenv("LW_SSH_USER") }, // THIS ROW MUST BE FIRST IN THE TABLE
```

Comment:
I would be tempted to move the `LW_SSH_USER` check between lines 153 and 154. Otherwise it could be really easy to introduce a regression by reordering the elements in this array and/or adding a new AMI type before the LW_SSH_USER check.

Reply:
I'd agree, but we test this case in a unit test. So I'd like to keep everything all neatly packaged up if possible